src/components/Widgets/PreviewHOC.js

```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+class PreviewHOC extends React.Component {
+
+  /**
+   * Only re-render on value change, but always re-render objects and lists.
+   * Their child widgets will each also be wrapped with this component, and
+   * will only be updated on value change.
+   */
+  shouldComponentUpdate(nextProps) {
+    const isWidgetContainer = ['object', 'list'].includes(nextProps.field.get('widget'));
+    return isWidgetContainer || this.props.value !== nextProps.value;
```

Comment:
Understanding this line requires remembering the precedence rules for `!==` and `||` - could we add some parens here?